File: PHP/DesignPatterns/Iterator.php
<?php


// Node class used for the Doubly Linked List.
class Node {
    
    public $value;
    public $next;
    public $prev;

    public function __construct($value, $next, $previous) { 
        $this->value = $value;
        $this->next = $next;
        $this->prev = $previous;
    }
}

// Simplified DLink:
class DoublyLinkedList {

    // Member variables for the access points of the list.
    private $head;
    private $tail;
    private $size;

    // Constructor
    public function __construct() {
        $this->head = null;
        $this->tail = null;
        $this->size = null;
    }

    public function isEmpty() {
        return $this->size <= 0;
    }

    public function size() {
        return $this->size;
    }

    // Inserts to the end of the list.
    public function addLast($value) {
        if($this->isEmpty()) {
            // If empty, insert and setup.
            $tmp = new Node($value, null, null);
            $this->head = $tmp;
            $this->tail = $tmp;
            $this->size++;
        }
        else 
        {
            $this->tail->next = new Node($value, null, $this->tail);
            $this->tail = $this->tail->next;
            $this->size++;
        }
    }

    public function getIterator(): DLinkIterator
    {
        return new DLinkIterator($this->head);
    }

    public  function getReverseIterator(): DLinkReverseIterator
    {
        return new DLinkReverseIterator($this->tail);
    }
}

class DLinkIterator
{
    protected $current;

    public function __construct($start)
    {
        $this->current = $start;
    }
    
    public function HasNext(): bool
    {
        return $this->current !== null;
    }

    public function Next()
    {
        if(!$this->HasNext())
        {
            throw new Exception("No more nodes to iterate!");
        }

        $val = $this->current->value;
        $this->current = $this->current->next;

        return $val;
    }
}

// Reverse iterator:
class DLinkReverseIterator
{
    protected $current;

    public function __construct($start)
    {
        $this->current = $start;
    }
    
    public function HasNext(): bool
    {
        return $this->current !== null;
    }

    public function Next()
    {
        if(!$this->HasNext())
        {
            throw new Exception("No more nodes to iterate!");
        }

        $val = $this->current->value;
        $this->current = $this->current->prev;

        return $val;
    }
}


// Demo:
$myList = new DoublyLinkedList();
$myList->addLast('Hello');
$myList->addLast('World');
$myList->addLast('Goodbye');
$myList->addLast('World');

echo "Forward iterator:\n";
$iter = $myList->getIterator();

while($iter->HasNext())
{
    echo $iter->Next() . "\n";
}

/*
    Output:
        Hello
        World
        Goodbye
        World
*/

echo "Reverse iterator:\n";
$revIter = $myList->getReverseIterator();

while($revIter->HasNext())
{
    echo $revIter->Next() . "\n";
}

/*
    Output:
        World
        Goodbye
        World
        Hello
*/

File: Python/DesignPatterns/Iterator.py
# Simplified version of Doubly Linked List:

class Node(object):
    def __init__(self, value, nextNode, prevNode):
        self.value = value
        self.next = nextNode
        self.prev = prevNode

class DoublyLinkedList(object):
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None
    
    def isEmpty(self):
        return self.head is None
    
    def addLast(self, value):
        if self.isEmpty():
            tmp = Node(value, None, None)
            self.head = tmp
            self.tail = tmp
            self.size += 1
            return
        
        self.tail.next = Node(value, None, self.tail)
        self.tail = self.tail.next
        self.size += 1

    def getIterator(self):
        return DLinkIterator(self.head)
    
    def getReverseIterator(self):
        return DLinkReverseIterator(self.tail)


# Forward Iterator:
class DLinkIterator(object):
    def __init__(self, startNode):
        self.current = startNode

    def hasNext(self):
        return self.current != None

    def next(self):
        if not self.hasNext():
            raise IndexError
        val = self.current.value
        self.current = self.current.next
        return val

# Backward Iterator:
class DLinkReverseIterator(object):
    def __init__(self, startNode):
        self.current = startNode

    def hasNext(self):
        return self.current != None

    def next(self):
        if not self.hasNext():
            raise IndexError
        val = self.current.value
        self.current = self.current.prev
        return val


# Demo:
myList = DoublyLinkedList()

myList.addLast("Hello")
myList.addLast("World")
myList.addLast("Goodbye")
myList.addLast("World")

# Iterate the list forwards:
forwardIter = myList.getIterator()

while forwardIter.hasNext():
    print(forwardIter.next())


'''
    Output:
        Hello
        World
        Goodbye
        World
'''

print("=======")

reverseIter = myList.getReverseIterator()

while reverseIter.hasNext():
    print(reverseIter.next())

'''
    Output:
        World
        Goodbye
        World
        Hello
'''